core/modules/wpublish/QueriesAndChunks.h

```diff
@@ -99,52 +103,47 @@ class ChunkStatsTable {
     std::uint64_t _tasksCompleted{0}; ///< Number of Tasks that have completed on this chunk/table.
     std::uint64_t _tasksBooted{0}; ///< Number of Tasks that have been booted for taking too long.
 
-    double _avgCompletionTime{0.0}; ///< weighted average of completion time
+    double _avgCompletionTime{0.0}; ///< weighted average of completion time in minutes.
     double _weightAvg{49.0}; ///< weight of previous average
-    double _weightDur{1.0}; ///< weight of new measurement
-    double _weightSum{_weightAvg + _weightDur}; ///< denominator
+    double _weightNew{1.0}; ///< weight of new measurement
+    double _weightSum{_weightAvg + _weightNew}; ///< denominator
 };
 
 
-/// Statistics for one chunk.
-class ChunkStats {
+/// Statistics for one chunk, including scan table statistics.
+class ChunkStatistics {
 public:
-    using Ptr = std::shared_ptr<ChunkStats>;
+    using Ptr = std::shared_ptr<ChunkStatistics>;
 
-    ChunkStats(int chunkId) : _chunkId{chunkId} {}
+    ChunkStatistics(int chunkId) : _chunkId{chunkId} {}
 
-    ChunkStatsTable::Ptr add(std::string const& scanTableName, double duration);
-    ChunkStatsTable::Ptr getStats(std::string const& scanTableName) const;
+    ChunkTableStats::Ptr add(std::string const& scanTableName, double duration);
+    ChunkTableStats::Ptr getStats(std::string const& scanTableName) const;
+
+    friend QueriesAndChunks;
 private:
     int const _chunkId;
     mutable std::mutex _tStatsMtx; ///< protects _tableStats;
     /// Map of chunk scan table statistics indexed by slowest scan table name in query.
-    std::map<std::string, ChunkStatsTable::Ptr> _tableStats;
+    std::map<std::string, ChunkTableStats::Ptr> _tableStats;
 };
 
 
-class QueryChunkStatistics {
+class QueriesAndChunks {
 public:
-    using Ptr = std::shared_ptr<QueryChunkStatistics>;
-
-    QueryChunkStatistics(std::chrono::seconds deadAfter) :_deadAfter{deadAfter} {
-        auto rDead = [this](){
-            while (_loopRemoval) {
-                removeDead();
-                std::this_thread::sleep_for(_deadAfter);
-            }
-        };
-        std::thread t(rDead);
-        _removalThread = move(t);
-    }
+    using Ptr = std::shared_ptr<QueriesAndChunks>;
 
-    virtual ~QueryChunkStatistics() {
+    QueriesAndChunks(std::chrono::seconds deadAfter,
+                         std::chrono::seconds examineAfter, int maxTasksBooted);
+    virtual ~QueriesAndChunks() {
         _loopRemoval = false;
+        _loopExamine = false;
         _removalThread.join();
+        _examineThread.join();
```

Comment:
`join()` could potentially throw, maybe worth catching all exceptions here (destructors should not throw).